File: projects/engineering.purdue.edu/kak/scriptingwithobjects/swocode/chap17/SOAPDaemonServer.pl
#!/usr/bin/perl -w

### SOAPDaemonServer.pl

use strict;
#use SOAP::Lite +trace;                                              #(A)
use SOAP::Transport::HTTP;                                           #(B)

# Don't want to die on 'Broken pipe' or Ctrl-C         
$SIG{PIPE} = $SIG{INT} = 'IGNORE';                                   #(C)

my $daemon = SOAP::Transport::HTTP::Daemon                           #(D)
              -> new( LocalAddr => 'localhost', LocalPort => 8080 )  #(E)
              -> dispatch_to( 'ACMEWebService' );                    #(F)

print "SOAP server started at ", $daemon->url, "\n";                 #(G)
$daemon->handle;                                                     #(H)

#---------  SOAP clients can call on the ACMEWebService class  ----------

package ACMEWebService;                                              #(I)

sub new {                                                            #(J)
    my $self = shift;                                                #(K)
    my $class = ref($self) || $self;                                 #(L)
    bless { _acemwebservice => shift } => $class;                    #(M)
}

sub hello {                                                          #(N)
    return "hello world from the soap daemon server";              
}

sub bye {                                                            #(O)
    return "goodbye, world from the soap daemon server";
}

sub SumAndDifference {                                               #(P)
    my ($self, $x, $y) = @_;
    return { "sum" => $x + $y, "difference" => $x - $y };
}

File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap17/SOAPDaemonServer.py
#!/usr/bin/env python

### SOAPDaemonServer.py

import SOAPpy                                                        #(A)

class ACMEWebService:                                                #(B)
    def hello(self):                                                 #(C)
       return "hello world from ACMEWebService " +  \
                 "of the stand-alone SOAP daemon server"             #(D)
    def bye(self):                                                   #(E)
        return "goodbye, world from the soap daemon server"          #(F)
    def SumAndDifference(self, arg1, arg2 ):                         #(G)
        return { "sum" : arg1 + arg2, "difference" : arg1 - arg2 }   #(H)

def sayHelloWorld():                                                 #(I)
    return "Hello World from a stand-alone SOAP Daemon Server"       #(J)

server = SOAPpy.SOAPServer(("localhost", 8080), \
                           namespace="urn:ACMEWebService")           #(K)

server.registerObject(ACMEWebService())                              #(L)
server.registerFunction(sayHelloWorld)                               #(M)
server.serve_forever()                                               #(N)
